add early async migration docs (#2596)
PostHog/posthog.com: contents/handbook/engineering/async-migrations.md

---
title: Async Migrations

> sidebar: Handbook

Author: should they live there later? this is too much noise for users to care about

## Why are async migrations necessary?

Migrations are inevitable, and sometimes it may be necessary to execute non-trivial schema changes that can take a long time to complete. 

For example, ClickHouse does not support changing the primary key of a table, which is a change we were [forced to make in anticipation of upgrading ClickHouse beyond version 21.6](https://github.com/PostHog/posthog/issues/5684). As a result, the way to change the schema of the table was to create a new table and insert all the data from the old table into it, which took us an entire week to run on PostHog Cloud.

Now, while we can execute such changes to PostHog Cloud "manually" (i.e. James G SSHs into CH5, runs a bunch of queries using clickhouse-client, and keeps an eye on them), we cannot expect the same of our users (plus this approach does not scale well for us either).

As a result, we needed to create a system capable of safely and efficiently managing migrations that need to happen asynchronously.

> ## Why did we write this system from scratch?

Author: I couldn't find anything open source that would do what we need out of the box. We could for example sit on top of the CH Infi ORM but that'd have been very limited for things we've added to migrations.

Another interesting thing is that GitLab also has an in-house solution, and run into the same issues with ee and non-ee code:
https://docs.gitlab.com/ee/development/background_migrations.html

They don't have a management UI though

As a provider of a product that users can self-host, we needed a bespoke solution that interoperates well with our particular tech stack and requirements, such as:

> 1. The ability to run asynchronous tasks using a system already in use (Celery)

Author: It kind of is. We're reluctant (rightfully so) to introduce new services at PostHog. It's in the handbook somewhere but can't find it rn. Adding another system for this when we already have a worker would be best to avoid although not impossible

2. Accounting for both databases we have in use (Postgres and ClickHouse) and their subsequent migrations
3. Accounting for our EE codebase
4. The ability for self-hosted users to manage and monitor migrations easily 
5. Accounting for our particular deployment options

> 6. Allowing logical branching (splitting code paths) in the future

Author: I'll just remove this section as it's noise

7. Extensibility to support future needs we might have for long-running operations

## How do async migrations work?

### Writing an async migration

To write an async migration, you should create a migration file inside `posthog/special_migrations/migrations`. The name should follow the convention we use for Django and EE migrations (e.g. `0005_update_events_schema`).

> Here's what an example async migration looks like:

Author: I wrote this new one specifically for this - as it's not a nice test/example in the repo but good for educational purposes. Not sure the best approach here. People are also less likely to visit the link than read inline stuff

> Here's what an example async migration looks like:

Reviewer: true, but it will get out of date (especially since it requires PRs against 2 repos), which I believe is more important. Alternatively you can mention that there are more examples and link to them, so even if it's out of date it's not that bad.

    def is_required(self):
        result = sync_execute("SELECT count(*) FROM system.tables WHERE database='posthog' AND name='table_x_new'")
        return result[0][0] == 0 

    # check if we're not running low on storage
    def healthcheck(self):
        result = sync_execute("SELECT free_space FROM system.disks")
        if result[0][0] < 1000000000:
            return (False, "ClickHouse available storage below 1GB") 

        return (True, None)

    # update progress based on the total rows moved
    def progress(self, _):
        result = sync_execute(f"SELECT count(*) FROM table_x")
        result2 = sync_execute(f"SELECT count(*) FROM table_x_new")
        total_rows_to_move = result2[0][0]
        total_rows_moved = result[0][0]

        progress = 100 * total_rows_moved / total_rows_to_move
        return progress

```

</details>

### Workflow and architecture 

#### Setup

When the Django server boots up - a setup step for async migrations happens, which does the following:

1. Imports all the migration definitions 
2. Populates a dependencies map and in-memory record of migration definitions
3. Creates a database record for each

> 4. Check if all async migrations necessary for this PostHog version have completed (else don't start)

Reviewer: As I called out in the PR this is a problem for the upgrade path specifically (https://github.com/PostHog/posthog/pull/7446#discussion_r765339085), we need to fail earlier here and not let folks get into a state where they tried an upgrade and errored out on special migrations. Please explain how we're going to solve that.

on that note - there are also customers who kubectl apply stuff (we don't really need to support them as that's very custom), but how can we make their experience best?

> 5. Triggers a migration to run if `AUTO_START_SPECIAL_MIGRATIONS` is set and there's an uncompleted migration for this version

Reviewer: ```suggestion
5. Triggers migrations to run (in order) if `AUTO_START_SPECIAL_MIGRATIONS` is set and there are uncompleted migrations for this version
```

#### Running a migration

When a migration is triggered, the following happens:

1. A task is dispatched to Celery to run this migration in the background
2. The following basic checks are performed to ensure the migration can indeed run:
    1. We're not over the concurrent migrations limit
    2. The migration can be run with the current PostHog version
    3. The migration is not already running
    4. The service version requirements are met (e.g. X < ClickHouse version < Y)
    5. The migration's `is_required` check passes
    6. The migration's `healthcheck` passes
    7. The migration's dependency (if any) has been completed
3. We run through each of the operations in order
4. Every 30 minutes, a Celery task performs a healthcheck, to ensure that:
   1. The Celery process running the migration didn't crash
   2. The migration's healthcheck still passes 

> **Note:** Async migrations can also be run synchronously (i.e. not in Celery) using the async migrations CLI (WIP) or the Django shell.

#### Stopping a migration

> A migration can be stopped by issuing a command via Celery's app control to terminate the process running the task. 

Reviewer: or via the cli, maybe give an example command here?

Author: Sorry I mean we do this in the background. The CLI doesn't exist yet :D

#### Rollbacks

> If a migration is stopped for any reason (manual trigger or error), we will attempt to roll back the migration following the operations specified in reverse order from the last started operation. 

Reviewer: how can I stop a migration without a rollback?

Author: you added a comment that I should add a flag for this and I will :D

If a roll back succeeds, the migration status will be updated to reflect this.

#### Errors

If a migration errors, the error message is added to the migration's database record and we automatically trigger a rollback.

### Scope and limitations

The initial implementation of async migrations targets only **data migrations**, and assumes that the migration is used as a mechanism to help users move into a new default state. 

For example, when we [moved our ClickHouse `person_distinct_id` table to a `CollapsingMergeTree`](https://github.com/PostHog/posthog/pull/5563), we updated the SQL for creating the table, and wrote a migration to help users on the old schema migrate to the new schema. 

However, users that did a fresh deploy of PostHog _after_ this change already had the table with the new schema created by default.

This is the only type of operation that async migrations _currently_ support, to prevent a complex web of dependencies between migration types.

As such, those writing an async migration should write a sensible `is_required` function that determines if the migration should run or not. 

> Thus, when a user deploys a new PostHog instance, we will first run **all** EE migrations in order, and then **all** of the async migrations in order. At this step, async migrations should be skipped if the codebase already contains updated default schemas. 

Reviewer: please explain how the codebase can already contain the updated default schemas, how does that work? Is the way that we'd write the same the special migration as a normal migration after the version upper bound is passed (via if table doesn't exist etc)?

Author: Take a look at the PR linked above. We updated the base SQL for the table and wrote a migration to get users on it. It's the normal way we do CH migrations actually.
